server/api/src/rsDataObjRepl.cpp

```diff
@@ -839,17 +914,65 @@ namespace
                 destination_replica.resc_hier(),
                 destination_replica.vote()));
 
-            if (irods::replication::is_allowed(_comm, source_replica, destination_replica, irods::replication::log_errors::no) &&
-                destination_replica.vote() > irv::vote::zero)
-            {
-                destination_cond_input[RESC_HIER_STR_KW] = destination_replica.resc_hier();
-                destination_cond_input[DEST_RESC_HIER_STR_KW] = destination_replica.resc_hier();
+            // If the source and destination hierarchies match, just skip this one.
+            if (source_replica.resc_id() == destination_replica.resc_id()) {
+                log_api::debug(
+                    "[{}:{}] - Destination hierarchy [{}] matches source hierarchy [{}]. Skipping replication.",
+                    __func__,
+                    __LINE__,
+                    destination_replica.resc_hier(),
+                    source_replica.resc_hier());
 
-                if (const int ec = replicate_data(_comm, source_inp, destination_inp, _stat);
-                    ec < 0 && status >= 0)
-                {
-                    status = ec;
-                }
+                continue;
+            }
+
+            log_api::info("src hier: [{}], src id: [{}], dest hier: [{}], dest id: [{}]",
+                          source_replica.resc_hier(),
+                          source_replica.resc_id(),
+                          destination_replica.resc_hier(),
+                          destination_replica.resc_id());
+
+            // If the destination replica did not vote positively, it should not be updated.
+            if (destination_replica.vote() <= irv::vote::zero) {
+                log_api::debug(
+                    "[{}:{}] - Selected destination replica in hierarchy [{}] is not available for update. Replication "
+                    "is not allowed.",
+                    __func__,
+                    __LINE__,
+                    destination_replica.resc_hier());
+
+                continue;
+            }
+
+            if (GOOD_REPLICA == destination_replica.replica_status()) {
+                // If the destination replica is good, it does not need to be updated. Skip it.
+                //
+                // TODO Should this trigger fileModified like regular replications do?
+                log_api::debug("[{}:{}] - Selected destination replica in hierarchy [{}] is already good. Skipping.",
+                               __func__,
+                               __LINE__,
+                               destination_replica.resc_hier());
+
+                continue;
+            }
```

Comment:
So this little block is running inside a loop which iterates over all replicas.

Do you see a situation where invoking fileModified seems correct?

Reply:
If the changes pass all of the tests, then I think we're fine. Until someone brings us a case we don't cover, I say let it ride.

Reply:
principle of tiny boxes says let the replication trigger fileModified... if for no other reason that it's easier for US to reason about?

Reply:
> Because this is going to update all the replicas anyway, I guess it doesn't need to invoke fileModified for replication. It may be needed for something like compound where you might expect sync-to-archive to trigger, though. Or some other future use case that leans on that operation.

Can you elaborate on the compound resource thought?

Reply:
We talked about this offline and have decided that it is more correct NOT to trigger fileModified in this case. `irepl -a` is surreptitiously circumventing policy - much like `iphymv` - and so any fileModified behavior will need to be achieved by a direct replication to the hierarchy or by invoking the desired operation through some other means. I will leave a comment to this effect and I created this issue for implementing proper testing around this in the future: https://github.com/irods/irods/issues/7493